src/pretty/to_fixed/mod.rs

```diff
@@ -0,0 +1,605 @@
+use crate::{
+    d2s::{DOUBLE_BIAS, DOUBLE_EXPONENT_BITS, DOUBLE_MANTISSA_BITS},
+    digit_table::DIGIT_TABLE,
+    pretty::{
+        format64,
+        to_fixed::d2fixed_full_table::{
+            ADDITIONAL_BITS_2, MIN_BLOCK_2, POW10_OFFSET, POW10_OFFSET_2, POW10_SPLIT,
+            POW10_SPLIT_2,
+        },
+    },
+};
+#[cfg(feature = "no-panic")]
+use no_panic::no_panic;
+
+mod d2fixed_full_table;
+
+/// Max bytes/characters required for `toFixed` representation of a [`f64`] value:
+///
+/// - 1 byte for sign (-)
+/// - `22` bytes for whole part:
+///     Because we have a check for if `>= 1e21` (1 byte extra, just in case)
+/// - `1` byte for dot (`.`)
+/// - `108` (`9 * 12`) bytes for fraction part:
+///     We write digits in blocks, which consist of `9` digits.
+///
+/// Total: `1 + 22 + 1 + 108 = 132`
+pub const MAX_BUFFER_SIZE: usize = 132;
+
+pub struct Cursor {
+    buffer: *mut u8,
+    len: isize,
+    index: isize,
+}
+
+impl Cursor {
+    pub fn new(buffer: *mut u8, len: usize) -> Self {
+        debug_assert!(!buffer.is_null());
+        Self {
+            buffer,
+            len: len as isize,
+            index: 0,
+        }
+    }
+
+    /// Append one byte to buffer.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that there is enough space for the given byte.
+    unsafe fn append_byte(&mut self, c: u8) {
+        debug_assert!(self.index < self.len);
+
+        *self.buffer.offset(self.index) = c;
+        self.index += 1;
+    }
+
+    /// Append the byte `count` times into the buffer.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that there is enough space for the given bytes.
+    unsafe fn append_bytes(&mut self, c: u8, count: usize) {
+        debug_assert!(self.index + count as isize <= self.len);
+
+        self.buffer.offset(self.index).write_bytes(c, count);
+        self.index += count as isize;
+    }
+
+    /// Gets the current [`Cursor`] index.
+    ///
+    /// The `index` is also the amount of bytes that have been written into the buffer.
+    fn index(&self) -> usize {
+        self.index as usize
+    }
+
+    /// Convert `digits` to decimal and write the last 9 decimal digits to result.
+    /// If `digits` contains additional digits, then those are silently ignored.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the buffer has enough space for `9` bytes.
+    unsafe fn append_nine_digits(&mut self, mut digits: u32) {
+        let count = 9;
+
+        debug_assert!(self.index + count <= self.len);
+
+        if digits == 0 {
+            self.append_bytes(b'0', 9);
+            return;
+        }
+
+        let result = self.buffer.offset(self.index);
+
+        for i in (0u32..5).step_by(4) {
```

Comment:
This should be the same right?

```suggestion
        for i in [0,4] {
```